x/hard/keeper/borrow_test.go

```diff
@@ -39,389 +38,580 @@ func (suite *KeeperTestSuite) TestBorrow() {
-						Expiry:        time.Now().Add(1 * time.Hour),
-					},
-					{
-						MarketID:      "kava:usd",
-						OracleAddress: sdk.AccAddress{},
-						Price:         tc.args.priceKAVA,
-						Expiry:        time.Now().Add(1 * time.Hour),
-					},
-					{
-						MarketID:      "btcb:usd",
-						OracleAddress: sdk.AccAddress{},
-						Price:         tc.args.priceBTCB,
-						Expiry:        time.Now().Add(1 * time.Hour),
-					},
-					{
-						MarketID:      "bnb:usd",
-						OracleAddress: sdk.AccAddress{},
-						Price:         tc.args.priceBNB,
-						Expiry:        time.Now().Add(1 * time.Hour),
-					},
-				},
-			}
-
-			// Initialize test application
-			tApp.InitializeFromGenesisStates(authGS,
-				app.GenesisState{pricefeedtypes.ModuleName: tApp.AppCodec().MustMarshalJSON(&pricefeedGS)},
-				app.GenesisState{types.ModuleName: tApp.AppCodec().MustMarshalJSON(&hardGS)})
-
-			// Mint coins to hard module account
-			bankKeeper := tApp.GetBankKeeper()
-			hardMaccCoins := sdk.NewCoins(sdk.NewCoin("ukava", sdkmath.NewInt(1000*KAVA_CF)),
-				sdk.NewCoin("usdx", sdkmath.NewInt(200*USDX_CF)), sdk.NewCoin("busd", sdkmath.NewInt(100*BUSD_CF)))
-			err := bankKeeper.MintCoins(ctx, types.ModuleAccountName, hardMaccCoins)
-			suite.Require().NoError(err)
-
-			keeper := tApp.GetHardKeeper()
-			suite.app = tApp
-			suite.ctx = ctx
-			suite.keeper = keeper
-
-			// Run BeginBlocker once to transition MoneyMarkets
-			hard.BeginBlocker(suite.ctx, suite.keeper)
-
-			err = suite.keeper.Deposit(suite.ctx, tc.args.borrower, tc.args.depositCoins)
-			suite.Require().NoError(err)
-
-			// Execute user's previous borrows
-			err = suite.keeper.Borrow(suite.ctx, tc.args.borrower, tc.args.previousBorrowCoins)
-			if tc.args.previousBorrowCoins.IsZero() {
-				suite.Require().True(strings.Contains(err.Error(), "cannot borrow zero coins"))
-			} else {
-				suite.Require().NoError(err)
-			}
-
-			// Now that our state is properly set up, execute the last borrow
-			err = suite.keeper.Borrow(suite.ctx, tc.args.borrower, tc.args.borrowCoins)
-
-			if tc.errArgs.expectPass {
-				suite.Require().NoError(err)
-
-				// Check borrower balance
-				acc := suite.getAccount(tc.args.borrower)
-				suite.Require().Equal(tc.args.expectedAccountBalance, suite.getAccountCoins(acc))
-
-				// Check module account balance
-				mAcc := suite.getModuleAccount(types.ModuleAccountName)
-				suite.Require().Equal(tc.args.expectedModAccountBalance, suite.getAccountCoins(mAcc))
-
-				// Check that borrow struct is in store
-				_, f := suite.keeper.GetBorrow(suite.ctx, tc.args.borrower)
-				suite.Require().True(f)
-			} else {
-				suite.Require().Error(err)
-				suite.Require().True(strings.Contains(err.Error(), tc.errArgs.contains))
+			for i, arg := range tc.args {
+				if i == 0 {
+					// Initialize test app and set context
+					tApp = app.NewTestApp()
+					ctx = tApp.NewContext(true, tmproto.Header{Height: 1, Time: tmtime.Now()})
```

Comment:
With the setup args moved to it's own field, we can then do the initialization before the loop without needing to check the index - only looping for borrows